clang/include/clang/CIR/Dialect/IR/CIRTypes.td

```diff
@@ -110,12 +110,22 @@ def CIR_StructType : CIR_Type<"Struct", "struct",
     "mlir::StringAttr":$typeName,
     "bool":$body,
     "bool":$packed,
-    "std::optional<::mlir::cir::ASTRecordDeclAttr>":$ast
+    "std::optional<::mlir::cir::ASTRecordDeclAttr>":$ast,
+    "mlir::cir::StructType::RecordKind":$kind
   );
 
   let hasCustomAssemblyFormat = 1;
 
   let extraClassDeclaration = [{
+    enum RecordKind : uint32_t {
+      CLASS,
+      UNION,
+      STRUCT,
+      // FIXME(cir): remove the kind below once we have the proper types for
+      // stuff like TypeInfo and VTableAttr.
+      NONE,
```

Comment:
Not sure I follow, both typeinfo and vtable are attributes with regular struct type, why would they need to be declared differently?

Reply:
My intention here is a separation of concerns.

We use `StructType` to represent not only structs but any record type in CIR. Stuff like `TypeInfo` and `VTableAttr` are represented as CIR-like records, but I don't think they should as there are attributes in `StructType` that will never be used by them (e.g. `ASTRecordDeclAttr`).

It's true that, if we were to have separate types for them, they would likely just be a subset of `StructType`, but I still think CIR would benefit from it from an organizational standpoint.

Reply:
> My intention here is a separation of concerns.
> 
> We use `StructType` to represent not only structs but any record type in CIR. Stuff like `TypeInfo` and `VTableAttr` are represented as CIR-like records,

Sometimes the compiler needs to synthesize types and whatnots as part of emitting information that's not written by the user, like in the case of vtables, rtti, more complex global initializers, etc - it's a normal fact that those will require use of existing types (like records, arrays) for making them available to runtime / linking purposes / etc.

> but I don't think they should as there are attributes in `StructType` that will never be used by them (e.g. `ASTRecordDeclAttr`).

I like to make them special at the attribute level, since they usually refer to some "constant" property, while re-using an existing type to represent them, which makes lowering direct and the only high level property we miss is the "attribute". `ASTRecordDeclAttr` is optional, and won't make much sense for things that the compiler has to synthesize, so I'm not sure it's a good example.

> It's true that, if we were to have separate types for them, they would likely just be a subset of `StructType`, but I still think CIR would benefit from it from an organizational standpoint.

Maybe, I really don't see how it's better than offloading the semantic to a special attribute of a said type though. Perhaps remove "NONE" for now and we can revisit it later if it makes sense to introduce?